File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.cpp
//{ Driver Code Starts
// Initial Template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int countSubarrays(vector<int> &nums, int k) {
        unordered_map<int, int> prefixSumFreq; // To store the frequency of prefix sums
        int currentSum = 0; // To keep track of the running prefix sum
        int count = 0;

        prefixSumFreq[0] = 1; // Base case: Subarray starting from index 0

        for (int num : nums) {
            currentSum += num; // Update the running sum
            
            // Check if (currentSum - k) exists in the map
            if (prefixSumFreq.find(currentSum - k) != prefixSumFreq.end()) {
                count += prefixSumFreq[currentSum - k];
            }

            // Update the frequency of the current prefix sum
            prefixSumFreq[currentSum]++;
        }

        return count;
    }
};


//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    cin.ignore();
    while (t--) {

        string ks;
        getline(cin, ks);
        int k = stoi(ks);
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        Solution obj;
        cout << obj.countSubarrays(arr, k);
        cout << endl;
        cout << "~"
             << "\n";
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.java
class Solution {
    public int cntSubarrays(int[] arr, int k) {
        // Create a HashMap to store prefix sums and their frequencies
        HashMap<Integer, Integer> map = new HashMap<>();
        int count = 0; // This will store the number of valid subarrays
        int sum = 0; // This keeps track of the running sum (prefix sum)

        // Initialize map with (0, 1) to handle cases where subarray starts from index 0
        // This means a sum of 0 has occurred once before we start
        map.put(0, 1); 

        // Iterate through each number in the array
        for (int num : arr) {
            // Add current number to the running sum
            sum += num;

            // Check if (sum - k) exists in the map
            // If it does, it means there are subarrays ending at current index
            // that sum to k
            if (map.containsKey(sum - k)) {
                // Add the number of times (sum - k) has occurred to count
                count += map.get(sum - k);
            }

            // Update the map with current sum
            // If sum already exists, increment its count
            // Otherwise, add it with count 1
            map.put(sum, map.getOrDefault(sum, 0) + 1);
        }

        return count;
    }
}